internal/pkg/gateway/registry.go

```diff
@@ -67,37 +73,61 @@ func (reg *registry) endorsers(channel string, chaincode string) ([]*endorser, e
 	// choose first layout for now - todo implement retry logic
 	if len(descriptor.Layouts) > 0 {
 		layout := descriptor.Layouts[0].QuantitiesByGroup
-		var r []*dp.Peer
+		var r []*endorserState
 		for group, quantity := range layout {
-			// For now, select n remoteEndorsers from each group at random. Future story, sort by block height
-			endorsers := e[group].Peers
-			rand.Shuffle(len(endorsers), func(i, j int) {
-				endorsers[i], endorsers[j] = endorsers[j], endorsers[i]
-			})
-			peerGroup := endorsers[0:quantity]
+			// Select n remoteEndorsers from each group sorted by block height
+
+			// block heights
+			var groupPeers []*endorserState
+			for _, peer := range e[group].Peers {
+				msg := &gossip.GossipMessage{}
+				err := proto.Unmarshal(peer.StateInfo.Payload, msg)
+				if err != nil {
+					return nil, err
```

Comment:
IIUC this actually changes the failure model of the function - we are now failing to select an endorser if we can't gather block height related information. Is that expected?

Reply:
If you can't discern channel related information about a peer, it will not show up in discovery queries anyway. 
So, all peers you get from discovery are those that have this information in the first place, assuming gossip doesn't misbehave.

Reply:
Right